### ISSUE TITLE
Error Creating Transfer After Upgrading from Version 0.2.4 to 1.0.6 – PostgreSQL ambiguous_function Operator Issue

### ISSUE THREAD

**Issue by zeadhani** [ash-project/ash_double_entry#66]
### **Issue Summary**

When creating a transfer after upgrading from version 0.2.4 to 1.0.6, an error occurs during the operation. The PostgreSQL error ambiguous_function appears, indicating that an operator is not unique, with the hint suggesting the need for explicit type casts.

### **Steps to Reproduce**

1. Upgrade ash_double_entry from version 0.2.4 to 1.0.6.
2. Run the transfer creation process using the following context:
      ->  Transfer involves inserting into or updating the balances table.
3. Encounter an error during the operation with the following debug logs:

`[debug] QUERY ERROR source="balances" db=0.0ms
UPDATE "balances" AS b0 SET "balance" = (CASE ... ) WHERE ...
** (Postgrex.Error) operator is not unique: - unknown
Hint: Could not choose a best candidate operator. You might need to add explicit type casts.
`

### **Observed Behavior**

The transfer creation fails with an ambiguous_function error in PostgreSQL.

### **Expected Behavior**

The transfer should be created without any errors, and the balances table should update correctly.

### **Additional Details**

- Elixir Version:  1.17.3
- Ash Framework Version: 3.0
- Postgrex.Error: Code 42725

### **Troubleshooting Done**

- Verified that no duplicate identifiers exist in accounts.
- Reviewed data types and confirmed they align with what is expected in the balances table.

![image](https://github.com/user-attachments/assets/e7776a7d-48b9-4dec-9a25-373e2806e10c)

**Comment by zachdaniel** [comment]
What specific ash, ash_postgres and ash_sql versions are you using?

**Comment by zachdaniel** [comment]
Also `ash_money`

**Comment by zachdaniel** [comment]
Please make sure you're on the latest of all of them.

**Comment by zeadhani** [comment]
I am using the following versions, all of which are updated:

	•	ash_postgres: ~> 2.4.12
	•	ash_sql: ~> 0.2.38
	•	ash_double_entry: ~> 1.0.6
	•	ash_money: ~> 0.1.13
	•	ex_money_sql: ~> 1.11

**Comment by zachdaniel** [comment]
I've made a small change in `main` that may help, but if not I will have to look into this next week. Please try it out and let me know.

**Comment by zeadhani** [comment]
Thank you for the quick response and update! I tried the change, but now I’m getting the following error:
`** (Postgrex.Error) ERROR 42725 (ambiguous_function) operator is not unique: + unknown`

Please let me know when it’s fixed or if there’s an issue on my end that I should address. Appreciate your help!

**Comment by zachdaniel** [comment]
🤔 something is definitely strange there. I'm not seeing the same error in my own application. Can you please create a reproduction? Have you applied the relevant `ash_money` migrations? I think there may be a set up issue somewhere along those lines.

**Comment by zeadhani** [comment]
I’ll double-check those migrations tomorrow morning and provide a quick update. 🙇‍♂️

**Comment by zachdaniel** [comment]
I actually found something that I think *may* be the cause. It requires a fix both in `Ash` and `AshMoney`. I will push the fix up to `main` of both and you can try it :)

It fixes a different issue that I noticed, but seems "adjacent" to this one :)

**Comment by zachdaniel** [comment]
Please give those packages (in `main`) a try and LMK. If not, a reproduction will be necessary.

**Comment by zeadhani** [comment]
Thanks for looking into this! 
I pulled all the packages from the main and re-ran all the migrations from scratch, but the error still exists. 

<img width="1001" alt="image" src="https://github.com/user-attachments/assets/c76f3eaa-3724-4fa0-b81f-97683c6ba066">

**Comment by zeadhani** [comment]
<img width="761" alt="image" src="https://github.com/user-attachments/assets/924de612-f191-4dbc-ba27-6ec14cfa6f76">

**Comment by zachdaniel** [comment]
Okay, please provide a small reproduction project so I can diagnose further

**Comment by zeadhani** [comment]
Will do

**Comment by zeadhani** [comment]
Here is a small reproduction project for you to diagnose further:

[[GitHub Repository Link](https://github.com/ash-project/ash_double_entry/issues/your-repo-link)](https://github.com/zeadhani/ash_double_entry_trial)

**Comment by zeadhani** [comment]
now I’m getting the following error:

<img width="960" alt="image" src="https://github.com/user-attachments/assets/c0af7d0e-30a1-43c9-aa6d-c3384b31e30c" />

**Comment by zeadhani** [comment]
The reproduction project I shared was built exactly as you described, using mix igniter.new example --install ash, ash_postgres, ash_money, and then setting up ash_double_entry from scratch. Please take a look when you get a chance.

Also, if you could check as soon as possible, that would be great, as this issue is blocking an urgent project. Let me know if you need any additional details!

**Comment by zachdaniel** [comment]
I will look as soon as my schedule allows but I cannot promise anything more :)

**Comment by zeadhani** [comment]
I appreciate it!
Thank you ;)

**Comment by zachdaniel** [comment]
I just cloned that project down and the tests all pass for me

**Comment by zeadhani** [comment]
I created two accounts and tried to make a transfer between them, but the issue appears on my side. Any idea what might be causing it or anything I should check?

**Comment by zachdaniel** [comment]
Could potentially be postgres/elixir version related? I'm on latest postgres and latest elixir versions 🤔 have you tried deleting your `_build` directory and doing a full fresh recompile?

**Comment by zeadhani** [comment]
Yup, I tried that. I’m on the latest Elixir version as well, but for Postgres, I’m using the latest Alpine image.

**Comment by zachdaniel** [comment]
🤔 This fails for you consistently in the same way? The same failure message every time you run `mix test`?

**Comment by zeadhani** [comment]
The issue isn’t with mix test—it happens when I create two accounts and attempt a transfer. The balance record fails to be created with a negative value on the debited account, causing the transaction and transfer to fail.

**Comment by zachdaniel** [comment]
Oh. Apologies, I misunderstood. Can you please add a test that reproduces the issue? I often spend time reproducing issues that people expect should be easy to reproduce and then I do something subtly different than them and it works for me etc. etc.

**Comment by zeadhani** [comment]
Thanks for your understanding! I’ll work on adding a test case to reproduce the issue and keep you updated. Appreciate your help! 🙏

**Comment by zeadhani** [comment]
I’ve added a test in the reproduction project—please pull the latest changes and check it out. The test fails, by the way. Let me know what you find! @zachdaniel

**Comment by zachdaniel** [comment]
In your reproduction project, the configuration:

```elixir
config :ash, :known_types, [AshMoney.Types.Money]
```

is missing. Adding it back in resolves the issue.

I spent 2 hours of my Sunday morning debugging this 😢 The installer command that I gave you sets this `known_types` configuration. I'm guessing the reproduction project was not generated fresh with those installers, otherwise your test would have passed and the issue would have been much more clear.

It may very well have been a bug with the installer the first time you ran it that has since been fixed. For this reason it is really important to follow these reproduction steps fresh when creating an issue, not assume that they were followed correctly in the past. Because it is my nights and weekends that pay the price ☹️ .

**Comment by zeadhani** [comment]
Thanks for taking the time to investigate this! I really appreciate it.
That makes sense—I’ll ensure a fresh setup next time. Appreciate your help! @zachdaniel
